### PR TITLE
Custom fetch for session expiration

### PR DESCRIPTION
**Pull request by narin**
## Description
Background: I’m working on a modal that will appear with a session expiration countdown that allows the user to extend their session.

The way we are keeping the user’s session expiration in sync with the backend is by extending the expiration time every time they make an API request.  If the user has a session, every API response will include a header `X-Session-Expiration`  with the updated expiration time.  The front end then stores this value in localStorage to keep track of if/when to display the modal.

This works fine for apps that are using our `apiRequest()`  utility method, which checks for the header.  However, there are about 25 instances in the app that are using `fetch()`  instead of the `apiRequest()`.

In an ideal world, I’d refactor all the code using `fetch()` to use `apiRequest()` instead, but Eugene and I found a lot of inconsistencies in the error handling with each instance doing something a little different.. You can probably see how much work and testing would be involved that.

This is an alternate solution where we create a custom fetch that checks the response headers if they are calling the API.  Any code currently `fetch()` should import this custom function instead.  

## Testing done
Tested locally.

## Acceptance criteria
- [ ] Custom fetch checks for `X-Session-Expiration` header
- [ ] All files using `fetch()` instead of `apiRequest()` should import this custom function

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: src/applications/facility-locator/api/LocatorApi.js [hunk 1]
```diff
@@ -1,5 +1,6 @@
 import compact from 'lodash/compact';
 import { api } from '../config';
+import { fetch } from '../../../platform/utilities/api';
```
Comment (ncksllvn) [inline src/applications/facility-locator/api/LocatorApi.js]: Not required, but you can actually omit the path to the directory and just lead with `platform/`. It makes imports across multiple files a little easier.

File: src/platform/utilities/api/index.js [hunk 2]
```diff
@@ -3,6 +3,33 @@ import * as Sentry from '@sentry/browser';
...
+        if (sessionExpiration) {
+          localStorage.setItem('sessionExpiration', sessionExpiration);
+        }
+      }
+      return response;
+    });
+  }
+
+  return global.fetch(...args);
+}
```
Comment (ncksllvn) [inline src/platform/utilities/api/index.js]: I feel a little weird about needing to reference the global fetch. Maybe we can name this function as something like `fetchAndUpdateSessionExpiration`, but then export it on a different line like `export { fetch: fetchAndUpdateSessionExpiration }`. Or, maybe we can export it as `fetchAndUpdateSessionExpiration` and import it in other modules like `import { fetchAndUpdateSessionExpiration as fetch } from '../../utilities/api'`. What do you think?

File: src/platform/utilities/api/index.js [hunk 3]
```diff
@@ -56,10 +83,6 @@ export function apiRequest(resource, optionalSettings = {}, success, error) {
         : Promise.resolve(response);
 
       if (response.ok || response.status === 304) {
-        // Get session expiration from header
-        const sessionExpiration = response.headers.get('X-Session-Expiration');
-        if (sessionExpiration)
-          localStorage.setItem('sessionExpiration', sessionExpiration);
```
Comment (ncksllvn) [inline src/platform/utilities/api/index.js]: Love that `apiRequest` leverages the local `fetch` function!

**Comment by rianfowler** [comment]
What if we polled this status instead spreading this behavior into every `fetch` call? The intent would be easier to see if it's localized to a single polling handler and we wouldn't burden people with knowing which APIs reset the session (and needing to know which fetch to use). 

I think we should endeavor to allow people to use what they're familiar with like `fetch` and work around that. Changing global functions incurs team debt / maintenance debt, enforcement costs, etc.

**Comment by ncksllvn** [comment]
@rianfowler @jbalboni actually mentioned that same idea when we were discussing a solution here. I think this is considered a good implementation for phase 1 due to time/scope, but polling would be a really good upgrade once there's more bandwidth.

I think have `fetch` as an import across files is okay, because it's also what we're during with browser APIs like `localStorage`. But maybe there should be a distinction between the `fetch` defined here and the global `fetch`?

**Comment by rianfowler** [comment]
Yeah I think if we want to have a decorated fetch we should name it differently and then make a ticket for VSP to follow up with the BE work needed to poll the status. We probably shouldn’t wrap global fetch or refer to the decorated version as `fetch`.

**Review by ncksllvn — APPROVED** [review]

**Pull request merged**